core/trino-main/src/main/java/io/trino/execution/StageStateMachine.java

```diff
@@ -638,6 +643,36 @@ public StageInfo getStageInfo(Supplier<Iterable<TaskInfo>> taskInfosSupplier)
                 failureInfo);
     }
 
+    private static List<OperatorStats> combineTaskOperatorSummaries(List<TaskInfo> taskInfos, int maxTaskOperatorSummaries)
+    {
+        // Group each unique pipelineId + operatorId combination into lists
+        Long2ObjectOpenHashMap<ArrayList<OperatorStats>> pipelineAndOperatorToStats = new Long2ObjectOpenHashMap<>(maxTaskOperatorSummaries);
+        // Expect to have one operator stats entry for each taskInfo
+        int taskInfoCount = taskInfos.size();
+        LongFunction<ArrayList<OperatorStats>> statsListCreator = key -> new ArrayList<>(taskInfoCount);
+        for (TaskInfo taskInfo : taskInfos) {
+            for (PipelineStats pipeline : taskInfo.getStats().getPipelines()) {
+                // Place the pipelineId in the high bits of the combinedKey mask
+                long pipelineKeyMask = Integer.toUnsignedLong(pipeline.getPipelineId()) << 32;
```

Comment:
nit: why not simply `((long) pipeline.getPipelineId()) << 32`? Or `(pipeline.getPipelineId() & 0xFFFFFFFFL) << 32` in case it ever for whatever reason becomes negative?

Reply:
Yeah, the handling of negatives (should never happen) was the motivation for `Integer.toUnsignedLong` for completeness. I think it makes more sense to call `toUnsignedLong` instead of using `(pipeline.getPipelineId() & 0xFFFFFFFFL) << 32`  since that's what the `toUnsignedLong` method already does internally and calling that method seems to make the intention more clear rather than having a large constant value of `0xFFFFFFFFL` directly in the code. I'm happy to change it though if you disagree.

Reply:
I don't have a strong opinion here. Somehow it feels like `toUnsignedLong` makes the intention less obvious. But I could be just me. Feel free to keep it as is.